### PR TITLE
Fix issue when trying to deploy with custom S3 bucket name (AWS)

### PR DESCRIPTION
**Pull request by HyperBrain**
## What did you implement:

References:
#3344 
#3463 

The new `generateCoreTemplate` module uses the `validateS3BucketName` function but does not include it anymore.

## How did you implement it:

Separated validateS3BucketName and validate. Separated unit tests. Added correct require to `generateCoreTemplate`.

### Important info
The _validate_ and _validateS3BucketName_ had to be split, as the _validate_ function must not be assigned to **this** again - this will override and clean all states for validate which would lead to failures in deploy afterwards.

## How can we verify it:

Set the `deploymentBucket` property in the serverless.yml file. Then use different combinations of `deploy`, `package` and `deploy --package=...`

I tested all three possible combinations - worked without problems.

## Todos:

- [x] Write tests
- [ ] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

**Comment by HyperBrain** [comment]
@pmuens Please verify on your side too. All package/deploy combinations worked for me.

@erikerikson I found the problem with your approach. It was, that "validate" was assigned twice to the package/deploy plugin and the settings were killed though. My solution was to separate validate() and validateS3BucketName(), so that validate is only applied by the new "aws:common:validate" lifecycle event.
Nevertheless, could you please test my fix too? On my system it worked with a custom bucket name set as `deploymentBucket` in the serverless.yml.

**Comment by eahefnawy** [comment]
@HyperBrain just tested this out with `deploymentBucket` in different cases. No issues found! thanks for the fix!

@erikerikson could you confirm that this PR fixes your issue so we can merge it? 😊

File: lib/plugins/aws/lib/validateS3BucketName.test.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
+'use strict';
+
+const expect = require('chai').expect;
+const Serverless = require('../../../Serverless');
+const validateS3BucketName = require('./validateS3BucketName');
+
+describe('#validate', () => {
```
Comment (pmuens) [inline lib/plugins/aws/lib/validateS3BucketName.test.js]: Should be updated to `#validateS3BucketName()`

File: lib/plugins/aws/package/lib/generateCoreTemplate.js [hunk 2]
```diff
@@ -4,8 +4,15 @@ const BbPromise = require('bluebird');
 const path = require('path');
 const _ = require('lodash');
 
+const validateS3BucketName = require('../../lib/validateS3BucketName');
+
 module.exports = {
   generateCoreTemplate() {
+    _.assign(
```
Comment (pmuens) [inline lib/plugins/aws/package/lib/generateCoreTemplate.js]: Could we add a test for this as well?

**Comment by erikerikson** [comment]
Reading the code, I trust that it solves the `validateS3BucketName` related issue.  No rocket science there.

It doesn't do anything about the choke on self-references issue that hits this bit of code:
https://github.com/serverless/serverless/blob/master/lib/classes/Utils.js#L51

If that's tracked elsewhere, apologies...  a bit consumed recently.

**Comment by erikerikson** [comment]
Thanks @HyperBrain!

**Comment by HyperBrain** [comment]
@pmuens I was not able to reproduce the circular JSON issue with the different test cases I ran. @erikerikson Can you provide some info about any "specialities" you used in your service configuration?

**Comment by HyperBrain** [comment]
Agree. It would be much easier to find the specific issue. Having it in `master` also will let it pop up if other users encounter it.

**Comment by erikerikson** [comment]
Did you test with a ${self:} value under custom?

**Comment by erikerikson** [comment]
Example: https://github.com/Nordstrom/hello-retail/blob/develop/retail-stream/serverless.yml#L12

**Comment by erikerikson** [comment]
For the record, I think it's a good idea to deal with that bug under a separate PR since it effects fewer users and doing so will resolve the more common blocked use case of custom deployment buckets but do please resolve it.

**Comment by HyperBrain** [comment]
@erikerikson Ok. Then let's merge it to master and you can create a new issue with your findings and samples.
@pmuens Can you assign the new issue to me afterwards? Then I will take care of the fix within the next few days.

**Comment by pmuens** [comment]
Thanks @HyperBrain and @erikerikson 👍 Appreciate it.

I'll open up another issue for that and test it today with `${self:}`.

**Comment by pmuens** [comment]
BTW. Just added the test case in the "Misc" section in #3467

**Pull request merged**